src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs

```diff
@@ -0,0 +1,192 @@
+using Microsoft.VisualStudio.Settings;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Options
+{
+    /// <summary>
+    /// Serializes settings marked with <see cref="RoamingProfilePersistence"/> to and from the user's roaming profile.
+    /// </summary>
+    [Export(typeof(IOptionPersister))]
+    internal sealed class RoamingVisualStudioProfileOptionPersister : ForegroundThreadAffinitizedObject, IOptionPersister
+    {
+        // NOTE: This service is not public or intended for use by teams/individuals outside of Microsoft. Any data stored is subject to deletion without warning.
+        [Guid("9B164E40-C3A2-4363-9BC5-EB4039DEF653")]
+        private class SVsSettingsPersistenceManager { };
+
+        private readonly ISettingsManager _settingManager;
+        private readonly IGlobalOptionService _globalOptionService;
+
+        /// <summary>
+        /// The list of options that have been been fetched from <see cref="_settingManager"/>, by key. We track this so
+        /// if a later change happens, we know to refresh that value. This is synchronized with monitor locks on
+        /// <see cref="_optionsToMonitorForChangesGate" />.
+        /// </summary>
+        private readonly Dictionary<string, List<OptionKey>> _optionsToMonitorForChanges = new Dictionary<string, List<OptionKey>>();
+        private readonly object _optionsToMonitorForChangesGate = new object();
+
+        /// <remarks>We make sure this code is from the UI by asking for all serializers on the UI thread in <see cref="HACK_AbstractCreateServicesOnUiThread"/>.</remarks>
+        [ImportingConstructor]
+        public RoamingVisualStudioProfileOptionPersister(IGlobalOptionService globalOptionService, [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+            : base(assertIsForeground: true) // The GetService call requires being on the UI thread or else it will marshal and risk deadlock
+        {
+            Contract.ThrowIfNull(globalOptionService);
+
+            this._settingManager = (ISettingsManager)serviceProvider.GetService(typeof(SVsSettingsPersistenceManager));
+            _globalOptionService = globalOptionService;
+
+            // While the settings persistence service should be available in all SKUs it is possible an ISO shell author has undefined the
+            // contributing package. In that case persistence of settings won't work (we don't bother with a backup solution for persistence
+            // as the scenario seems exceedingly unlikely), but we shouldn't crash the IDE.
+            if (this._settingManager != null)
+            {
+                ISettingsSubset settingsSubset = this._settingManager.GetSubset("*");
+                settingsSubset.SettingChangedAsync += OnSettingChangedAsync;
+            }
+        }
+
+        private System.Threading.Tasks.Task OnSettingChangedAsync(object sender, PropertyChangedEventArgs args)
```

Comment:
this is private and task returning, but returns EmptyTask?  WHy not just make this void and non-async?